### PR TITLE
feat(obstacle_stop_planner): improve chattering prevention mechanism

### PR DESCRIPTION
**Pull request by satoshi-ota**
### Related Link

- https://github.com/tier4/autoware_launch/pull/711

### Before

- the module only keep the info that whether there are obstacel points or not.
- chattering stop position

https://user-images.githubusercontent.com/44889564/212215832-f1df0055-5611-4a6e-b66a-5a89438feb05.mp4

### After

- the module also keep the history of the obstacle points' position and detection time.
- not chattering stop position

https://user-images.githubusercontent.com/44889564/212215844-6e4f11da-c04a-4abb-bd6d-179a8332f0a2.mp4

### TEST

Please set hunting_threshold to positive value (> 0).

```yaml
hunting_threshold: 0.5                   # even if the obstacle disappears, the stop judgment continues for hunting_threshold [s]
```

https://github.com/autowarefoundation/autoware_launch/blob/9ef2b6939d8aa276c909a6513f401ffc4301cce6/autoware_launch/config/planning/scenario_planning/lane_driving/motion_planning/obstacle_stop_planner/obstacle_stop_planner.param.yaml#L3

### Evaluator Result

[[TIER IV INTERNAL LINK]](https://evaluation.tier4.jp/evaluation/reports/90c8f428-9792-5046-a937-fd20a83ff28e?project_id=prd_jt) This PR(822/825)
[[TIER IV INTERNAL LINK]](https://evaluation.tier4.jp/evaluation/reports/c5dd33f7-9767-5955-843d-fd4de758ab1a?project_id=prd_jt) Baseline(820/825)

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by taikitanaka3** [comment]
@h-ohta 
could you review this PR?

File: planning/obstacle_stop_planner/include/obstacle_stop_planner/node.hpp [hunk 1]
```diff
@@ -121,14 +122,12 @@ class ObstacleStopPlannerNode : public rclcpp::Node
 
   std::unique_ptr<AdaptiveCruiseController> acc_controller_;
   std::shared_ptr<ObstacleStopPlannerDebugNode> debug_ptr_;
-  boost::optional<StopPoint> latest_stop_point_{boost::none};
   boost::optional<SlowDownSection> latest_slow_down_section_{boost::none};
+  std::vector<std::pair<rclcpp::Time, pcl::PointXYZ>> obstacle_history_{};
```
Comment (h-ohta) [inline planning/obstacle_stop_planner/include/obstacle_stop_planner/node.hpp:126]: making struct would be reasonable and easy to understand.

```
struct ObstacleHistory
{
  rclcpp::Time time
  pcl::PointXYZ point
}
```

Comment (satoshi-ota) [reply planning/obstacle_stop_planner/include/obstacle_stop_planner/node.hpp]: fixed in https://github.com/autowarefoundation/autoware.universe/pull/2647/commits/39ad6e23e50e8608a3bb2dabd956beabfa310650

File: planning/obstacle_stop_planner/include/obstacle_stop_planner/node.hpp [hunk 2]
```diff
@@ -91,6 +92,17 @@ using vehicle_info_util::VehicleInfo;
 using TrajectoryPoints = std::vector<TrajectoryPoint>;
 using PointCloud = pcl::PointCloud<pcl::PointXYZ>;
 
+struct ObstacleWithDetectionTime
+{
+  ObstacleWithDetectionTime() = default;
+  ObstacleWithDetectionTime(const rclcpp::Time & t, pcl::PointXYZ & p) : detection_time(t), point(p)
```
Comment (h-ohta) [inline planning/obstacle_stop_planner/include/obstacle_stop_planner/node.hpp:98]: Do you have any reason why you use default?
The following would be better.
```suggestion
  explicit ObstacleWithDetectionTime(const rclcpp::Time & t, pcl::PointXYZ & p) : detection_time(t), point(p)
```

Comment (satoshi-ota) [reply planning/obstacle_stop_planner/include/obstacle_stop_planner/node.hpp:98]: Thanks for your advise. There is no clear reason :sweat_smile: I'll fix as you said.

Comment (satoshi-ota) [reply planning/obstacle_stop_planner/include/obstacle_stop_planner/node.hpp]: fixed in https://github.com/autowarefoundation/autoware.universe/pull/2647/commits/02b7c18c86ec056ec11856ddee140fc21653db68

**Review by h-ohta — APPROVED** [review]

**Review by taikitanaka3 — APPROVED** [review]

File: planning/obstacle_stop_planner/include/obstacle_stop_planner/node.hpp [hunk 3]
```diff
@@ -202,6 +212,31 @@ class ObstacleStopPlannerNode : public rclcpp::Node
...
+      if (expired) {
+        itr = obstacle_history_.erase(itr);
+        continue;
+      }
+
+      itr++;
+    }
+  }
+
+  PointCloud::Ptr getOldPointCloudPtr() const
```
Comment (sfukuta) [inline planning/obstacle_stop_planner/include/obstacle_stop_planner/node.hpp:230]: `PointCloud`が定義がないためビルドエラーになる。
[autowarefoundation/autoware.universe/#1823](https://github.com/autowarefoundation/autoware.universe/pull/1823)にて、定義の設定と宣言を全体的に修正を実施。

**Pull request merged**